File: jserver-security/src/test/java/com/github/gr3gdev/jserver/security/http/TokenRequestTest.java
package com.github.gr3gdev.jserver.security.http;

import com.github.gr3gdev.jserver.http.Request;
import com.github.gr3gdev.jserver.security.TokenClientPlugin;
import com.github.gr3gdev.jserver.security.TokenServerPlugin;
import com.github.gr3gdev.jserver.security.user.UserData;
import org.bouncycastle.asn1.x509.ExtendedKeyUsage;
import org.bouncycastle.asn1.x509.KeyPurposeId;
import org.bouncycastle.asn1.x509.X509Extensions;
import org.bouncycastle.asn1.x509.X509Name;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.x509.X509V3CertificateGenerator;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import javax.security.auth.x500.X500Principal;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.math.BigInteger;
import java.security.*;
import java.security.cert.X509Certificate;
import java.util.Date;
import java.util.Optional;

import static org.junit.Assert.*;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class TokenRequestTest {

    @Mock
    private Request request;

    private File keystoreFile = new File("build/test.jks");
    private String keystorePassword = "TestPassword";

    private PublicKey publicKey;
    private PrivateKey privateKey;

    private static class UserTest implements UserData {

        private String name;
        private Integer count;

        public UserTest() {
            this.name = null;
            this.count = null;
        }

        public UserTest(String name, int count) {
            this.name = name;
            this.count = count;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public Integer getCount() {
            return count;
        }

        public void setCount(Integer count) {
            this.count = count;
        }
    }

    private X509Certificate generateSelfSignedCertificate(KeyPair keyPair) throws Exception {
        X509V3CertificateGenerator certGen = new X509V3CertificateGenerator();
        X500Principal dnName = new X500Principal("cn=example");
        certGen.setSerialNumber(BigInteger.valueOf(System.currentTimeMillis()));
        certGen.setSubjectDN(new X509Name("dc=name"));
        certGen.setIssuerDN(dnName);
        certGen.setNotBefore(new Date(System.currentTimeMillis() - 24 * 60 * 60 * 1000));
        certGen.setNotAfter(new Date(System.currentTimeMillis() + 2L * 365 * 24 * 60 * 60 * 1000));
        certGen.setPublicKey(keyPair.getPublic());
        certGen.setSignatureAlgorithm("SHA256WithRSAEncryption");
        certGen.addExtension(X509Extensions.ExtendedKeyUsage, true,
                new ExtendedKeyUsage(KeyPurposeId.id_kp_timeStamping));
        return certGen.generate(keyPair.getPrivate(), "BC");
    }

    @Before
    public void before() throws Exception {
        Security.addProvider(new BouncyCastleProvider());

        KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
        ks.load(null, keystorePassword.toCharArray());

        KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance("RSA");
        keyPairGenerator.initialize(2048);
        KeyPair keyPair = keyPairGenerator.generateKeyPair();

        publicKey = keyPair.getPublic();
        privateKey = keyPair.getPrivate();

        X509Certificate[] certificateChain = new X509Certificate[1];
        certificateChain[0] = generateSelfSignedCertificate(keyPair);
        ks.setKeyEntry("test", keyPair.getPrivate(), keystorePassword.toCharArray(), certificateChain);

        try (FileOutputStream fos = new FileOutputStream(keystoreFile)) {
            ks.store(fos, keystorePassword.toCharArray());
        }
    }

    @After
    public void after() {
        if (keystoreFile.exists()) {
            assertTrue(keystoreFile.delete());
        }
    }

    @Test
    public void testGetTokenFromHeaderMissing() {
        when(request.headers(TokenRequest.AUTH)).thenReturn(Optional.empty());
        TokenRequest.getTokenFromHeader(request).ifPresent(h -> fail("Token present"));
        assertTrue(true);
    }

    @Test
    public void testGetTokenFromHeaderBasic() {
        when(request.headers(TokenRequest.AUTH)).thenReturn(Optional.of("Basic ABCDEFGHIJKL"));
        TokenRequest.getTokenFromHeader(request).ifPresent(h -> fail("Token present"));
        assertTrue(true);
    }

    @Test
    public void testGetTokenFromHeader() {
        final String token = String.valueOf(new SecureRandom().nextLong());
        when(request.headers(TokenRequest.AUTH)).thenReturn(Optional.of("Bearer " + token));
        TokenRequest.getTokenFromHeader(request)
                .ifPresentOrElse(it -> assertEquals(token, it),
                        () -> fail("Token missing"));
    }

    @Test
    public void testGetTokenFromCookie1() {
        final String token = String.valueOf(new SecureRandom().nextLong());
        when(request.headers(TokenRequest.COOKIES)).thenReturn(Optional.of("COOKIE1=A; MY_COOKIE1=" + token + "; COOKIE2=B; COOKIE3=C"));
        TokenRequest.getTokenFromCookie(request, "MY_COOKIE1")
                .ifPresentOrElse(it -> assertEquals(token, it),
                        () -> fail("Cookie not found"));
    }

    @Test
    public void testGetTokenFromCookie2() {
        final String token = String.valueOf(new SecureRandom().nextLong());
        when(request.headers(TokenRequest.COOKIES)).thenReturn(Optional.of("COOKIE1=ABCDEFGHIJKL; MY_COOKIE2=" + token));
        TokenRequest.getTokenFromCookie(request, "MY_COOKIE2")
                .ifPresentOrElse(it -> assertEquals(token, it),
                        () -> fail("Cookie not found"));
    }

    @Test
    public void testGetUserDataFromToken() throws IOException {
        TokenServerPlugin tokenServer = new TokenServerPlugin(privateKey)
                .id("jServer").issuer("GR3Gdev");
        TokenClientPlugin tokenClient = new TokenClientPlugin(publicKey);
        UserTest userData = new UserTest("Name1", 1);
        String token = tokenServer.createToken(userData, (long) (60 * 60 * 1000));
        tokenClient.getUserData(token, UserTest.class).ifPresentOrElse(it -> {
            UserTest data = it.getData();
            assertEquals("Name1", data.name);
            assertEquals(1, data.count.intValue());
        }, () -> fail("UserData not found"));
    }
}

File: jserver-security/src/test/kotlin/com/github/gr3gdev/jserver/security/http/TokenRequestTest.kt
@file:Suppress("JAVA_MODULE_DOES_NOT_EXPORT_PACKAGE")

package com.github.gr3gdev.jserver.security.http

import com.github.gr3gdev.jserver.http.RemoteAddress
import com.github.gr3gdev.jserver.http.Request
import com.github.gr3gdev.jserver.logger.Logger
import com.github.gr3gdev.jserver.security.TokenClientPlugin
import com.github.gr3gdev.jserver.security.TokenServerPlugin
import com.github.gr3gdev.jserver.security.user.UserData
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import sun.security.x509.*
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.math.BigInteger
import java.security.*
import java.security.cert.X509Certificate
import java.util.*
import kotlin.collections.HashMap


class TokenRequestTest {

    private val keystoreFile = "build/keystore.jks"
    private val keystorePassword = "Secret4Keystore"
    private lateinit var publicKey: PublicKey

    private class UserTest : UserData {

        var name: String?
        var count: Int?

        constructor() {
            this.name = null
            this.count = null
        }

        constructor(name: String, count: Int) {
            this.name = name
            this.count = count
        }
    }

    private class RequestTest(val headers: Map<String, String>) : Request {
        override fun path(): String = ""
        override fun method(): String = "GET"
        override fun protocol(): String = "http"
        override fun headers(key: String): Optional<String> = Optional.ofNullable(headers[key])
        override fun <T> headers(key: String, ifPresent: (header: String) -> T, orElse: () -> T): T {
            var res: T? = null
            headers(key).ifPresentOrElse({
                res = ifPresent(it)
            }, {
                res = orElse()
            })
            return res!!
        }

        override fun headers(key: String, value: String) {}
        override fun headersNames(): MutableSet<String> {
            TODO("Not yet implemented")
        }

        override fun params(key: String): Optional<String> {
            TODO("Not yet implemented")
        }

        override fun <T> params(key: String, ifPresent: (param: String) -> T, orElse: () -> T): T {
            TODO("Not yet implemented")
        }

        override fun params(key: String, value: String) {
            TODO("Not yet implemented")
        }

        override fun paramsNames(): MutableSet<String> {
            TODO("Not yet implemented")
        }

        override fun remoteAddress(): RemoteAddress {
            TODO("Not yet implemented")
        }
    }

    @Before
    fun `Configure logger`() {
        Logger.changeLevel(Logger.Level.DEBUG)
    }

    private fun generateCertificate(keyPair: KeyPair): X509Certificate {
        val privateKey = keyPair.private
        val info = X509CertInfo()
        val from = Date()
        val to = Date(from.time + 5 * 1000L * 24L * 60L * 60L)
        val interval = CertificateValidity(from, to)
        val serialNumber = BigInteger(64, SecureRandom())
        val owner = X500Name("cn=Unknown")
        var sigAlgId = AlgorithmId(AlgorithmId.MD5_oid)
        info[X509CertInfo.VALIDITY] = interval
        info[X509CertInfo.SERIAL_NUMBER] = CertificateSerialNumber(serialNumber)
        info[X509CertInfo.SUBJECT] = owner
        info[X509CertInfo.ISSUER] = owner
        info[X509CertInfo.KEY] = CertificateX509Key(keyPair.public)
        info[X509CertInfo.VERSION] = CertificateVersion(CertificateVersion.V3)
        info[X509CertInfo.ALGORITHM_ID] = CertificateAlgorithmId(sigAlgId)

        // Sign the cert to identify the algorithm that's used.
        var certificate = X509CertImpl(info)
        certificate.sign(privateKey, "SHA256withRSA")

        // Update the algorith, and resign.
        sigAlgId = certificate[X509CertImpl.SIG_ALG] as AlgorithmId
        info[CertificateAlgorithmId.NAME + "." + CertificateAlgorithmId.ALGORITHM] = sigAlgId
        certificate = X509CertImpl(info)
        certificate.sign(privateKey, "SHA256withRSA")
        return certificate
    }

    @Before
    fun `Init keystore`() {
        val file = File(keystoreFile)
        val keystore = KeyStore.getInstance(KeyStore.getDefaultType())
        val keyGen = KeyPairGenerator.getInstance("RSA")
        keyGen.initialize(1024)
        val keyPair = keyGen.generateKeyPair()
        this.publicKey = keyPair.public

        val chain = arrayOf(generateCertificate(keyPair))

        keystore.load(null, keystorePassword.toCharArray())
        keystore.setKeyEntry("test", keyPair.private, keystorePassword.toCharArray(), chain)
        FileOutputStream(file).use {
            keystore.store(it, keystorePassword.toCharArray())
        }
    }

    @After
    fun `Delete keystore`() {
        File(keystoreFile).delete()
    }

    @Test
    fun `Test getToken from header missing`() {
        val request = RequestTest(HashMap())
        TokenRequest.getTokenFromHeader(request, {
            fail("Token present")
        }, {
            assertTrue(true)
        })
    }

    @Test
    fun `Test getToken from header Basic`() {
        val request = RequestTest(mapOf(Pair("Authorization", "Basic ABCDEFGHIJKL")))
        TokenRequest.getTokenFromHeader(request, {
            fail("Token present")
        }, {
            assertTrue(true)
        })
    }

    @Test
    fun `Test getToken from header`() {
        val token = SecureRandom().nextLong().toString()
        val request = RequestTest(mapOf(Pair("Authorization", "Bearer $token")))
        TokenRequest.getTokenFromHeader(request, {
            assertEquals(token, it)
        }, {
            fail("Token missing")
        })
    }

    @Test
    fun `Test getToken from Cookie 1`() {
        val token = SecureRandom().nextLong().toString()
        val request = RequestTest(mapOf(Pair("Cookie", "COOKIE1=A; MY_COOKIE1=$token; COOKIE2=B; COOKIE3=C")))
        TokenRequest.getTokenFromCookie(request, "MY_COOKIE1", {
            assertEquals(token, it)
        }, {
            fail("Cookie not found")
        })
    }

    @Test
    fun `Test getToken from Cookie 2`() {
        val token = SecureRandom().nextLong().toString()
        val request = RequestTest(mapOf(Pair("Cookie", "COOKIE1=ABCDEFGHIJKL; MY_COOKIE2=$token")))
        TokenRequest.getTokenFromCookie(request, "MY_COOKIE2", {
            assertEquals(token, it)
        }, {
            fail("Cookie not found")
        })
    }

    @Test
    fun `Test getUserData from token`() {
        val tokenServer = TokenServerPlugin("test", FileInputStream(keystoreFile), keystorePassword.toCharArray())
                .id("jServer").issuer("GR3Gdev")
        val tokenClient = TokenClientPlugin(publicKey)
        val userData = UserTest("Name1", 1)
        val token = tokenServer.createToken(userData, 60 * 60 * 1000)
        tokenClient.getUserData(token, UserTest::class.java, {
            val data = it.data
            assertEquals("Name1", data.name)
            assertEquals(1, data.count)
        }, {
            fail("UserData not found")
        })
    }
}
